### DISCUSSION TITLE
mpd_oled development version

### DISCUSSION THREAD

**Discussion by antiprism** [rern/rAudio#162] [General]
EDIT:

**The mpd_oled support for rAudio has been released and the development code is no longer available. Please see thread https://github.com/rern/rAudio-1/discussions/211 for the announcement, and for any further discussion about mpd_oled on rAudio**

ORIGINAL MESSAGE:
Hi

mpd_oled is an OLED status and spectrum display for Raspberry Pi.

   https://github.com/antiprism/mpd_oled/

I have been updating the build recently, in a development branch, and have added support for rAudio

   https://github.com/antiprism/mpd_oled/tree/dev

![mpd_oled_raudio](https://user-images.githubusercontent.com/2940060/106727309-a1ca5f80-660b-11eb-91cb-82a4ba062a05.jpg)

If anyone would like to try this. Here are the instructions

   https://github.com/antiprism/mpd_oled/blob/dev/INSTALL_RAUDIO.md

IMPORTANT: to download the development code when building mpd_oled, do not run `git clone https://github.com/antiprism/mpd_oled`, instead run
```
git clone -b dev https://github.com/antiprism/mpd_oled.git
```

Adrian.

**Comment by BV-Labs** [comment]
Cool :+1:

**Comment by Losiek68** [comment]
Great! I will give it a try. Just bought some SPI OLED displays.

**Reply by antiprism** [reply]
Please let me know how it goes, and whether anything on the instructions could be clearer.

**Reply by Losiek68** [reply]
Sure, I will. But I bought them China's Aliexpress so delivery will take some time...

**Reply by Losiek68** [reply]
Hi, I have just run my OLED display with mpd_oled. I had to install mpd_oled and cava from sources (I have Raspberry Pi 4 and you don't support Pi 4). It runs fine without any issues.
Now I will try to use polish diacritics and (maybe) remove spectrum analyzer.
Thank you for mpd_oled!

**Reply by antiprism** [reply]
Hi

Thanks for letting me know how the install went.

The binary package should support the Pi 4. I have a Pi 4 without an OLED attached running Raspberry Pi OS (Debian), and have just checked the package binary and it does it run on the Pi 4.

I am in the processes of switching to U8G2 for the display controllers and graphics, and this includes a number of fonts, so I may soon be able to add better language support to mpd_oled (rather than simply stripping accents).

Adrian.

**Reply by Losiek68** [reply]
Hi,
I got:
```
2021-03-08 13:15:46 (2.18 MB/s) - 'mpd-oled-0.2-1-armv6h.pkg.tar.zst' saved [551214/551214]

loading packages...
error: failed to prepare transaction (package architecture is not valid)
:: package mpd-oled-0.2-1-armv6h does not have a valid architecture
```

when running:
`sudo bash mpd_oled_raudio_install_latest.sh`
so I thought that Pi 4 is not supported. Anyway it runs like a charm.

**Reply by antiprism** [reply]
Hi

Thanks for posting the error messages. It looks like the packaging system of the Pi 4 rAudio image doesn't know about the armv6h architecture, and therefore that the mpd_oled binary package probably only installs on the Pi Zero image.

If the mpd_oled armv6h binary will run on the Pi4 image, I don't know whether the armv6h architecture could be enabled with a simple setting just to say it is fine, or whether this would pull in armv6h versions of lots of packages. Anyway, for now, I'll just document that the mpd_oled Arch package will only install on a Pi Zero.

Thanks again for letting me know.

Adrian.

**Comment by rern** [comment]
This should be a good addition to Addons.

Adrian, please provide a complete installation in a single script.

**Reply by Losiek68** [reply]
Great! But it is worth to mention that using I2C display you cannot use Power Button feature. That's why I bought SPI displays.

**Reply by antiprism** [reply]
Hi rern

I am in the process of streamlining the installation in order to provide Debian packages, which do a full install, but will always require configuring afterwards (to set the OLED type). A single package works on Volumio (Debian Jessie) and Moode (Debian Buster) and so the contained binaries may well work on rAudio, and maybe the Debian package could be converted into an Arch package using a converter like debtap.

Configuring MPD to make a copy of the audio for the spectrum analyser is easy on rAudio, but it would be even easier if the base /etc/mpd.conf had an `include_optional` line (RuneAudio's /etc/mpd.conf includes /home/your-extra-mpd.conf, but maybe the file should be somewhere unde /etc). I could then then script this part of the configuration.

I raised a feature request for a copy of the playing audio here:  https://github.com/rern/rAudio-1/discussions/163

Adrian.

**Reply by Losiek68** [reply]
Do you think about some customization? I don't like spectrum analyzers and I want to remove one and use released space for text.

**Reply by antiprism** [reply]
Hi Losiek68

The project has been in "maintenance mode" for quite a while, but I have started developing again and this is the current development plan: https://github.com/antiprism/mpd_oled/issues/52#issuecomment-774987783 . If you know some C++ it is not hard to amend the mpd_oled code to make a custom layout without the spectrum, see: https://github.com/antiprism/mpd_oled/blob/dev/doc/FAQ.md

Adrian.

**Reply by Losiek68** [reply]
Thank you, I'm not a coder but I wrote some scripts in Python, PowerShell, Bash. I don't know C++, but I will try to do some mods trying to learn C++. Maybe I'll be able to do it :)

**Reply by rern** [reply]
@antiprism,

`mpd_oled_FIFO`:
- `install.sh`
```sh
sed -i '/mpdconf+=\$btoutput/ a\
#oled0\
audio_output {\
	type           "fifo"\
	name           "mpd_oled_FIFO"\
	path           "/tmp/mpd_oled_fifo"\
	format         "44100:16:2"\
}\
#oled1
' /srv/http/bash/mpd-conf.sh
```
- `uninstall_oled.sh`
```sh
sed -i '/#oled0/,/#oled1/ d' /srv/http/bash/mpd-conf.sh
```

**Reply by rern** [reply]
> Great! But it is worth to mention that using I2C display you cannot use Power Button feature. That's why I bought SPI displays.

If you can solder it to the pins, both GPIO TFT LCD and Power Button will still work just fine.

**Reply by antiprism** [reply]
Hi rern

Thanks for looking at this and for providing the scripts. I don't know how updates occur on rAudio, but will changing /srv/http/bash/mpd-conf.sh block any updates, or if not, is there any chance this file will be overwritten by an update?

Adrian.

**Reply by rern** [reply]
Hi Adrian,

Once you've done with the installation script, I'll include the lines, with a conditional check, to the file itself.

rern

**Reply by Losiek68** [reply]
> > Great! But it is worth to mention that using I2C display you cannot use Power Button feature. That's why I bought SPI displays.
> 
> If you can solder it to the pins, both GPIO TFT LCD and Power Button will still work just fine.

You are right, my comment is obsolete now when there are two pins used to turn off Pi. My comment was related to one pin power off version.

**Reply by antiprism** [reply]
Hi rern

Here are the install instructions: https://gist.github.com/antiprism/44100b6c2b688f5d15a5c25172793f4b

The install and uninstall sections give the commands that would be run in a script (they require no user confirmations).

The commands to configure and remove the mpd.conf section are bundled in the package (I had to change the sed command for the install)
https://github.com/antiprism/mpd_oled/blob/dev2/scripts/mpd_oled_raudio_mpd_conf_install
https://github.com/antiprism/mpd_oled/blob/dev2/scripts/mpd_oled_raudio_mpd_conf_uninstall

I will wait to see how you include the mpd.conf configuration, but my usual instructions do not require changing a system file, and it would be good if there was no conflict with these instructions or, alternatively, if the addon method was also available for people not using the addon
https://github.com/antiprism/mpd_oled/blob/dev2/doc/install_raudio1_zst.md#configure-a-copy-of-the-playing-audio
Another possibility is for the changes in the usual instructions to be scripted and used for the mpd_oled addon (with the configuration file saved under /etc rather than /home which I would also change in the instructions), rather than changing a system file.

I made an arch package from my deb package using debtap, which required modifying: https://github.com/helixarch/debtap/issues/30 . The binaries should work on all Pis, but I have only tested on a Pi Zero.

After running the install commands the mpd_oled service needs to be configured with a command like 'mpd_oled_service_edit -o 6 -b 10 -g 1 -f 20`

I probably won't update the binary package for a while unless there is a major issue.

Adrian.

**Reply by rern** [reply]
Hi Adrian,
  
I have setup a repository for mpd_oled addon [here](https://github.com/rern/mpd_oled):
- Update rAudio 1
- While at Addons page: `sed -i '/i2c/ {n;n;n;d}' /srv/http/data/addons/addons-list.json`
- Refresh the page to show `MPD OLED Display` addon
- Try Install / Uninstall.

Once you've got familiar with it:
- Fork the repo to your GitHub page.
- Modify and Pull Request.

rern

**Reply by antiprism** [reply]
Hi rern

I don't use rAudio apart from checking compatbility with mpd_oled, so I would prefer to leave further work on the addon for someone that knows the system better than me.

For now, the Arch package I made is fairly convenient, and the user might prefer to edit the boot configuration files themself, particularly if they are using more than one I2C device.

I am about to incorporate the support for rAudio into the mpd_oled master branch. When this is done I will start a new discussion thread here to announce the availability of this, and to deal with rAudio-specific issues. I will also ask if anyone would like to be involved with preparing and maintaining the addon.

Adrian.

**Reply by rern** [reply]
Hi Adrian,

I've setup a `PKGBUILD` script for your [`mpd_oled` on AUR](https://aur.archlinux.org/packages/mpd_oled/) so it can be installed normally by `pacman`.

Build and install:
```sh
# build
su alarm
cd
curl -L https://aur.archlinux.org/cgit/aur.git/snapshot/mpd_oled.tar.gz | bsdtar xf -
cd mpd_oled
makepkg

# install
su
pacman -U mpd_oled*
```
Let me know your AUR username so I can transfer it to you.

rern

**Reply by rern** [reply]
**Spectrum OLED** is now a feature of **rAudio**.
- System > GPIO Deviecs > Spectrum OLED
- Fullscreen spectrum visualizer only (A stripped-down version of `mpd_oled` by Adrian)

**Reply by antiprism** [reply]
Hi rern

Sounds good, and thanks for the interest, but I am not able to maintain an Arch package at the moment.

If it is of inteest, I am in the process of switching mpd_oled to use the U8g2 library for controllers and graphics, so it works on more screens and has better font support

   [https://github.com/antiprism/mpd_oled_dev](https://github.com/antiprism/mpd_oled_dev)

 However, progress is slow...

Adrian.
